This notebook contains the different models and their accuracy using EBPSO model.

# DATA LOADING


In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('data_up.csv')

In [3]:
data.head(5)

SEQN  Gender  Age  Annual-Family-Income  Ratio-Family-Income-Poverty  \
0     2       1   77                     8                         5.00   
1     5       1   49                    11                         5.00   
2    12       1   37                    11                         4.93   
3    13       1   70                     3                         1.07   
4    14       1   81                     5                         2.67   

   X60-sec-pulse  Systolic  Diastolic  Weight  Height  ...  Total-Cholesterol  \
0             68        98         56    75.4   174.0  ...               5.56   
1             66       122         83    92.5   178.3  ...               7.21   
2             64       174         99    99.2   180.0  ...               4.03   
3            102       130         66    63.6   157.7  ...               8.12   
4             72       136         61    75.5   166.2  ...               4.50   

    HDL  Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
0  1.39              4.7              3              3                 1   
1  1.08              5.5              1              1                 1   
2  0.98              5.2              2              1                 1   
3  1.28              7.6              3              3                 1   
4  1.04              5.8              1              1                 1   

   Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  CoronaryHeartDisease  
0         2                   2                 2                     0  
1         2                   2                 2                     0  
2         2                   1                 1                     0  
3         1                   1                 2                     0  
4         2                   2                 2                     0  

[5 rows x 51 columns]

In [4]:
data.columns

Index(['SEQN', 'Gender', 'Age', 'Annual-Family-Income',
       'Ratio-Family-Income-Poverty', 'X60-sec-pulse', 'Systolic', 'Diastolic',
       'Weight', 'Height', 'Body-Mass-Index', 'White-Blood-Cells',
       'Lymphocyte', 'Monocyte', 'Eosinophils', 'Basophils', 'Red-Blood-Cells',
       'Hemoglobin', 'Mean-Cell-Vol', 'Mean-Cell-Hgb-Conc.',
       'Mean-cell-Hemoglobin', 'Platelet-count', 'Mean-Platelet-Vol',
       'Segmented-Neutrophils', 'Hematocrit', 'Red-Cell-Distribution-Width',
       'Albumin', 'ALP', 'AST', 'ALT', 'Cholesterol', 'Creatinine', 'Glucose',
       'GGT', 'Iron', 'LDH', 'Phosphorus', 'Bilirubin', 'Protein', 'Uric.Acid',
       'Triglycerides', 'Total-Cholesterol', 'HDL', 'Glycohemoglobin',
       'Vigorous-work', 'Moderate-work', 'Health-Insurance', 'Diabetes',
       'Blood-Rel-Diabetes', 'Blood-Rel-Stroke', 'CoronaryHeartDisease'],
      dtype='object')

In [5]:
data.shape

(37079, 51)

In [6]:
data.isnull().sum()

SEQN                           0
Gender                         0
Age                            0
Annual-Family-Income           0
Ratio-Family-Income-Poverty    0
X60-sec-pulse                  0
Systolic                       0
Diastolic                      0
Weight                         0
Height                         0
Body-Mass-Index                0
White-Blood-Cells              0
Lymphocyte                     0
Monocyte                       0
Eosinophils                    0
Basophils                      0
Red-Blood-Cells                0
Hemoglobin                     0
Mean-Cell-Vol                  0
Mean-Cell-Hgb-Conc.            0
Mean-cell-Hemoglobin           0
Platelet-count                 0
Mean-Platelet-Vol              0
Segmented-Neutrophils          0
Hematocrit                     0
Red-Cell-Distribution-Width    0
Albumin                        0
ALP                            0
AST                            0
ALT                            0
Cholesterol                    0
Creatinine                     0
Glucose                        0
GGT                            0
Iron                           0
LDH                            0
Phosphorus                     0
Bilirubin                      0
Protein                        0
Uric.Acid                      0
Triglycerides                  0
Total-Cholesterol              0
HDL                            0
Glycohemoglobin                0
Vigorous-work                  0
Moderate-work                  0
Health-Insurance               0
Diabetes                       0
Blood-Rel-Diabetes             0
Blood-Rel-Stroke               0
CoronaryHeartDisease           0
dtype: int64

Conclusion : Data is clean and all the data is numerical
CoronaryHeartDisease is target class. This is a classification problem predicting the heart disease.
SEQN is just representing the row number hence we need not take in the prediction columns.
Here, we have 49 features and 1 target column.
We have originally 37079 rows with no cell being empty.

# DATA ANALYSIS


In [7]:
data.describe()

SEQN        Gender           Age  Annual-Family-Income  \
count  37079.000000  37079.000000  37079.000000          37079.000000   
mean   48901.041236      1.513282     48.943661              7.358208   
std    26753.636441      0.499830     18.010440              3.994083   
min        2.000000      1.000000     20.000000              1.000000   
25%    26120.500000      1.000000     33.000000              4.000000   
50%    50065.000000      2.000000     48.000000              7.000000   
75%    71173.500000      2.000000     63.000000             10.000000   
max    93702.000000      2.000000     85.000000             15.000000   

       Ratio-Family-Income-Poverty  X60-sec-pulse      Systolic     Diastolic  \
count                 37079.000000   37079.000000  37079.000000  37079.000000   
mean                      2.559026      72.579250    124.090078     69.919253   
std                       1.624789      12.242108     19.254741     13.575804   
min                       0.000000      32.000000      0.000000      0.000000   
25%                       1.140000      64.000000    111.000000     62.000000   
50%                       2.180000      72.000000    121.000000     70.000000   
75%                       4.130000      80.000000    134.000000     78.000000   
max                       5.000000     224.000000    270.000000    132.000000   

             Weight        Height  ...  Total-Cholesterol           HDL  \
count  37079.000000  37079.000000  ...       37079.000000  37079.000000   
mean      80.988276    167.389601  ...           5.081713      1.370344   
std       20.678734     10.122908  ...           1.072682      0.415985   
min       32.300000    129.700000  ...           1.530000      0.160000   
25%       66.500000    160.000000  ...           4.320000      1.070000   
50%       78.200000    167.100000  ...           5.020000      1.290000   
75%       92.100000    174.600000  ...           5.740000      1.600000   
max      371.000000    204.500000  ...          14.090000      5.840000   

       Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
count     37079.000000   37079.000000   37079.000000      37079.000000   
mean          5.676496       1.783840       1.598856          1.218587   
std           1.050223       0.448324       0.511199          0.461102   
min           2.000000       1.000000       1.000000          1.000000   
25%           5.200000       2.000000       1.000000          1.000000   
50%           5.400000       2.000000       2.000000          1.000000   
75%           5.800000       2.000000       2.000000          1.000000   
max          18.800000       3.000000       3.000000          9.000000   

           Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  \
count  37079.000000        37079.000000      37079.000000   
mean       1.907333            1.549502          1.796165   
std        0.349674            0.497550          0.402853   
min        1.000000            1.000000          1.000000   
25%        2.000000            1.000000          2.000000   
50%        2.000000            2.000000          2.000000   
75%        2.000000            2.000000          2.000000   
max        3.000000            2.000000          2.000000   

       CoronaryHeartDisease  
count          37079.000000  
mean               0.040670  
std                0.197527  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                1.000000  

[8 rows x 51 columns]

# Feature Selection


**ALGORITHM : EBPSO**

Installing MealPy

In [27]:
!pip install mealpy --quiet

Importing necessary libraries

In [28]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from mealpy.swarm_based.PSO import C_PSO
from mealpy.utils.space import BinaryVar

Loading Dataset

In [29]:
df = pd.read_csv('data_up.csv')

In [30]:
df.head(5)

SEQN  Gender  Age  Annual-Family-Income  Ratio-Family-Income-Poverty  \
0     2       1   77                     8                         5.00   
1     5       1   49                    11                         5.00   
2    12       1   37                    11                         4.93   
3    13       1   70                     3                         1.07   
4    14       1   81                     5                         2.67   

   X60-sec-pulse  Systolic  Diastolic  Weight  Height  ...  Total-Cholesterol  \
0             68        98         56    75.4   174.0  ...               5.56   
1             66       122         83    92.5   178.3  ...               7.21   
2             64       174         99    99.2   180.0  ...               4.03   
3            102       130         66    63.6   157.7  ...               8.12   
4             72       136         61    75.5   166.2  ...               4.50   

    HDL  Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
0  1.39              4.7              3              3                 1   
1  1.08              5.5              1              1                 1   
2  0.98              5.2              2              1                 1   
3  1.28              7.6              3              3                 1   
4  1.04              5.8              1              1                 1   

   Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  CoronaryHeartDisease  
0         2                   2                 2                     0  
1         2                   2                 2                     0  
2         2                   1                 1                     0  
3         1                   1                 2                     0  
4         2                   2                 2                     0  

[5 rows x 51 columns]

In [31]:
df.columns

Index(['SEQN', 'Gender', 'Age', 'Annual-Family-Income',
       'Ratio-Family-Income-Poverty', 'X60-sec-pulse', 'Systolic', 'Diastolic',
       'Weight', 'Height', 'Body-Mass-Index', 'White-Blood-Cells',
       'Lymphocyte', 'Monocyte', 'Eosinophils', 'Basophils', 'Red-Blood-Cells',
       'Hemoglobin', 'Mean-Cell-Vol', 'Mean-Cell-Hgb-Conc.',
       'Mean-cell-Hemoglobin', 'Platelet-count', 'Mean-Platelet-Vol',
       'Segmented-Neutrophils', 'Hematocrit', 'Red-Cell-Distribution-Width',
       'Albumin', 'ALP', 'AST', 'ALT', 'Cholesterol', 'Creatinine', 'Glucose',
       'GGT', 'Iron', 'LDH', 'Phosphorus', 'Bilirubin', 'Protein', 'Uric.Acid',
       'Triglycerides', 'Total-Cholesterol', 'HDL', 'Glycohemoglobin',
       'Vigorous-work', 'Moderate-work', 'Health-Insurance', 'Diabetes',
       'Blood-Rel-Diabetes', 'Blood-Rel-Stroke', 'CoronaryHeartDisease'],
      dtype='object')

In [32]:
X = df.drop(columns=['SEQN', 'CoronaryHeartDisease'])

In [33]:
X.columns

Index(['Gender', 'Age', 'Annual-Family-Income', 'Ratio-Family-Income-Poverty',
       'X60-sec-pulse', 'Systolic', 'Diastolic', 'Weight', 'Height',
       'Body-Mass-Index', 'White-Blood-Cells', 'Lymphocyte', 'Monocyte',
       'Eosinophils', 'Basophils', 'Red-Blood-Cells', 'Hemoglobin',
       'Mean-Cell-Vol', 'Mean-Cell-Hgb-Conc.', 'Mean-cell-Hemoglobin',
       'Platelet-count', 'Mean-Platelet-Vol', 'Segmented-Neutrophils',
       'Hematocrit', 'Red-Cell-Distribution-Width', 'Albumin', 'ALP', 'AST',
       'ALT', 'Cholesterol', 'Creatinine', 'Glucose', 'GGT', 'Iron', 'LDH',
       'Phosphorus', 'Bilirubin', 'Protein', 'Uric.Acid', 'Triglycerides',
       'Total-Cholesterol', 'HDL', 'Glycohemoglobin', 'Vigorous-work',
       'Moderate-work', 'Health-Insurance', 'Diabetes', 'Blood-Rel-Diabetes',
       'Blood-Rel-Stroke'],
      dtype='object')

Ensuring x is a DataFrame and y is 1D array

In [34]:
y = df['CoronaryHeartDisease']

Defining the fitness Function and Train/test split

In [36]:
def fitness(sol):
    mask = sol > 0.5
    if not mask.any():
        return 1
    Xs = X.loc[:, mask]

    Xtr, Xte, ytr, yte = train_test_split(Xs, y, test_size=0.2, random_state=42)
    clf = RandomForestClassifier(random_state=42).fit(Xtr, ytr)

    return 1 - accuracy_score(yte, clf.predict(Xte))

Defining Binary Variables

In [37]:
bounds = [BinaryVar() for _ in range(X.shape[1])]
problem = {
    "obj_func": fitness,
    "bounds": bounds,
    "minmax": "min",
}

Run Binary

In [38]:
model = C_PSO(epoch=5, pop_size=20)
result = model.solve(problem)

INFO:mealpy.swarm_based.PSO.C_PSO:C_PSO(epoch=5, pop_size=20, c1=2.05, c2=2.05, w_min=0.4, w_max=0.9)
INFO:mealpy.swarm_based.PSO.C_PSO:>>>Problem: P, Epoch: 1, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 255.66917 seconds
INFO:mealpy.swarm_based.PSO.C_PSO:>>>Problem: P, Epoch: 2, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 267.83469 seconds
INFO:mealpy.swarm_based.PSO.C_PSO:>>>Problem: P, Epoch: 3, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 253.45055 seconds
/usr/local/lib/python3.12/dist-packages/mealpy/swarm_based/PSO.py:545: RuntimeWarning: invalid value encountered in scalar divide
  temp1 = self.w_min + (self.w_max - self.w_min) * (fit - fit_min) / (fit_avg - fit_min)
INFO:mealpy.swarm_based.PSO.C_PSO:>>>Problem: P, Epoch: 4, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 113.14981 seconds
/usr/local/lib/python3.12/dist-packages/mealpy/swarm_based/

Extracting Best Solution

In [48]:
best_position = result.solution
best_fitness = getattr(result, "solution_fitness", None) or getattr(result, "fitness_history", [None])[-1]

Selected Features

In [49]:
selected_features = np.where(best_position > 0.5)[0]
X_selected = X.iloc[:, selected_features] if selected_features.size > 0 else pd.DataFrame()

Calculating Accuracy and printing Results


In [50]:
if X_selected.shape[1] > 0:
    clf_final = RandomForestClassifier(random_state=42)
    accuracy = np.mean(cross_val_score(clf_final, X_selected, y, cv=5))
else:
    accuracy = 0

print("Best Fitness (1 - Accuracy):", best_fitness)
print("Selected Features Indices:", selected_features)
print("Total Features Selected:", len(selected_features))
print("Accuracy of Selected Features:", accuracy)

Best Fitness (1 - Accuracy): None
Selected Features Indices: [ 0  2  3  4  5  6  8  9 13 15 19 20 25 26 29 30 32 34 38 43 44 45 47 48]
Total Features Selected: 24
Accuracy of Selected Features: 0.9593031123680753


# MODEL SELECTION


In [51]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB

models = {
    'RandomForestClassifier' : RandomForestClassifier(random_state=42),
    'LogisticRegression' : LogisticRegression(random_state=42),
    'SupportVectorMachine' : SVC(random_state=42),
    'K-NearestNeighbors' : KNeighborsClassifier(),
    'DecisionTree' : DecisionTreeClassifier(random_state=42),
    'GaussianNaiveBayes' : GaussianNB()
}

In [52]:
X = df.drop(columns=['SEQN', 'CoronaryHeartDisease'])
y = df['CoronaryHeartDisease']
filtered_X = X.iloc[:, selected_features]

In [53]:
for name, model in models.items():
    X_train, X_test, y_train, y_test = train_test_split(filtered_X, y, test_size=0.2, random_state=42)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"{name} Accuracy: {accuracy}")

RandomForestClassifier Accuracy: 0.9586030204962244


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression Accuracy: 0.9571197411003236
SupportVectorMachine Accuracy: 0.9583333333333334
K-NearestNeighbors Accuracy: 0.9564455231930961
DecisionTree Accuracy: 0.92138619201726
GaussianNaiveBayes Accuracy: 0.938915857605178
